Bio/Cluster/mysort.h

```diff
@@ -0,0 +1,407 @@
+#pragma once
+#include <string.h>
+#include <limits.h>
+
+/*
+* C qsort() is very slow, much slower than C++ std::sort().
+* This is because qsort() doesn't utilize data-type information at compile time,
+* and it has redundant pointer dereference since it requires a compare function.
+* For projects that use old C, it's impossible to convert to C++/newer C.
+* 
+* So we implement a simple quicksort that is ~~25% faster than std::sort() 
+* with mostly random data, and much faster with structured/sorted data 
+*/
+
+static const int INF = INT_MAX; // 2^31 - 1
+
+/*	we use macro instead of function to avoid extra pointer dereference. */
+static double TEMP_SWAP_F64;
+#define swap_f64(x,y) {TEMP_SWAP_F64 = (x); (x) = (y); (y) = TEMP_SWAP_F64;}
+
+static int TEMP_SWAP_INT;
+#define swap_int(x,y) {TEMP_SWAP_INT = (x); (x) = (y); (y) = TEMP_SWAP_INT;}
+
+/* For quicksort, we need to choose a random pivot. Any random function should work. Even bad ones.
+   cycle size = 25000004
+*/
+static int
+cheap_random()
+{
+	const int base = 2 * 100 * 1000 * 1000 + 33;
+	static int seed = 0;
+	seed = seed * 7 + 13;
+	if (seed > base) seed %= base;
+	return seed;
+}
+
+/* Another way to choose pivot is using median of left, middle, and right values*/
+static inline int
+median_index_of3(const double arr[], const int a, const int b, const int c)
+{
+	if (arr[a] < arr[b]) {
+		if (arr[b] < arr[c]) return b;
+		else if (arr[a] < arr[c]) return c;
+		else return a;
+	}
+	else {
+		if (arr[a] < arr[c]) return a;
+		else if (arr[b] < arr[c]) return c;
+		else return b;
+	}
+}
+
+//****************************
+//****************************	MAIN SORTING PART
+
+/* Insertion sort is best when the array is small*/
+void
```

Comment:
Can all functions be declared `static`?

Reply:
Yes. But shouldn't the 2 functions that the user call directly (`fastsort` and` fastsort_index`) be normal and not `static` so that it can be called from other files?

Originally I intended to make all the other functions (except those two) `static` so that the file could be separated into a `.h` and `.c` file.

Reply:
To have a clean interface, I think we should declare fastsort and fastsort_index also `static` by integrating the new sorting code into cluster.c.

Reply:
I have put the content of mysort.h into cluster.c . But why is codecov/project showing errors in files that I didn't touch?